File: 2020/day10.js
// --- Day 10: Adapter Array ---

// Patched into the aircraft's data port, you discover weather forecasts of a massive tropical storm. Before you can figure out whether it will impact your vacation plans, however, your device suddenly turns off!

// Its battery is dead.

// You'll need to plug it in. There's only one problem: the charging outlet near your seat produces the wrong number of jolts. Always prepared, you make a list of all of the joltage adapters in your bag.

// Each of your joltage adapters is rated for a specific output joltage (your puzzle input). Any given adapter can take an input 1, 2, or 3 jolts lower than its rating and still produce its rated output joltage.

// In addition, your device has a built-in joltage adapter rated for 3 jolts higher than the highest-rated adapter in your bag. (If your adapter list were 3, 9, and 6, your device's built-in adapter would be rated for 12 jolts.)

// Treat the charging outlet near your seat as having an effective joltage rating of 0.

// Since you have some time to kill, you might as well test all of your adapters. Wouldn't want to get to your resort and realize you can't even charge your device!

// If you use every adapter in your bag at once, what is the distribution of joltage differences between the charging outlet, the adapters, and your device?

// For example, suppose that in your bag, you have adapters with the following joltage ratings:

// 16
// 10
// 15
// 5
// 1
// 11
// 7
// 19
// 6
// 12
// 4

// With these adapters, your device's built-in joltage adapter would be rated for 19 + 3 = 22 jolts, 3 higher than the highest-rated adapter.

// Because adapters can only connect to a source 1-3 jolts lower than its rating, in order to use every adapter, you'd need to choose them like this:

// The charging outlet has an effective rating of 0 jolts, so the only adapters that could connect to it directly would need to have a joltage rating of 1, 2, or 3 jolts. Of these, only one you have is an adapter rated 1 jolt (difference of 1).
// From your 1-jolt rated adapter, the only choice is your 4-jolt rated adapter (difference of 3).
// From the 4-jolt rated adapter, the adapters rated 5, 6, or 7 are valid choices. However, in order to not skip any adapters, you have to pick the adapter rated 5 jolts (difference of 1).
// Similarly, the next choices would need to be the adapter rated 6 and then the adapter rated 7 (with difference of 1 and 1).
// The only adapter that works with the 7-jolt rated adapter is the one rated 10 jolts (difference of 3).
// From 10, the choices are 11 or 12; choose 11 (difference of 1) and then 12 (difference of 1).
// After 12, only valid adapter has a rating of 15 (difference of 3), then 16 (difference of 1), then 19 (difference of 3).
// Finally, your device's built-in adapter is always 3 higher than the highest adapter, so its rating is 22 jolts (always a difference of 3).
// In this example, when using every adapter, there are 7 differences of 1 jolt and 5 differences of 3 jolts.

// Here is a larger example:

// 28
// 33
// 18
// 42
// 31
// 14
// 46
// 20
// 48
// 47
// 24
// 23
// 49
// 45
// 19
// 38
// 39
// 11
// 1
// 32
// 25
// 35
// 8
// 17
// 7
// 9
// 4
// 2
// 34
// 10
// 3

// In this larger example, in a chain that uses all of the adapters, there are 22 differences of 1 jolt and 10 differences of 3 jolts.

// Find a chain that uses all of your adapters to connect the charging outlet to your device's built-in adapter and count the joltage differences between the charging outlet, the adapters, and your device. What is the number of 1-jolt differences multiplied by the number of 3-jolt differences?

// --- Part Two ---

// To completely determine whether you have enough adapters, you'll need to figure out how many different ways they can be arranged. Every arrangement needs to connect the charging outlet to your device. The previous rules about when adapters can successfully connect still apply.

// The first example above (the one that starts with 16, 10, 15) supports the following arrangements:

// (0), 1, 4, 5, 6, 7, 10, 11, 12, 15, 16, 19, (22)
// (0), 1, 4, 5, 6, 7, 10, 12, 15, 16, 19, (22)
// (0), 1, 4, 5, 7, 10, 11, 12, 15, 16, 19, (22)
// (0), 1, 4, 5, 7, 10, 12, 15, 16, 19, (22)
// (0), 1, 4, 6, 7, 10, 11, 12, 15, 16, 19, (22)
// (0), 1, 4, 6, 7, 10, 12, 15, 16, 19, (22)
// (0), 1, 4, 7, 10, 11, 12, 15, 16, 19, (22)
// (0), 1, 4, 7, 10, 12, 15, 16, 19, (22)

// (The charging outlet and your device's built-in adapter are shown in parentheses.) Given the adapters from the first example, the total number of arrangements that connect the charging outlet to your device is 8.

// The second example above (the one that starts with 28, 33, 18) has many arrangements. Here are a few:

// (0), 1, 2, 3, 4, 7, 8, 9, 10, 11, 14, 17, 18, 19, 20, 23, 24, 25, 28, 31,
// 32, 33, 34, 35, 38, 39, 42, 45, 46, 47, 48, 49, (52)

// (0), 1, 2, 3, 4, 7, 8, 9, 10, 11, 14, 17, 18, 19, 20, 23, 24, 25, 28, 31,
// 32, 33, 34, 35, 38, 39, 42, 45, 46, 47, 49, (52)

// (0), 1, 2, 3, 4, 7, 8, 9, 10, 11, 14, 17, 18, 19, 20, 23, 24, 25, 28, 31,
// 32, 33, 34, 35, 38, 39, 42, 45, 46, 48, 49, (52)

// (0), 1, 2, 3, 4, 7, 8, 9, 10, 11, 14, 17, 18, 19, 20, 23, 24, 25, 28, 31,
// 32, 33, 34, 35, 38, 39, 42, 45, 46, 49, (52)

// (0), 1, 2, 3, 4, 7, 8, 9, 10, 11, 14, 17, 18, 19, 20, 23, 24, 25, 28, 31,
// 32, 33, 34, 35, 38, 39, 42, 45, 47, 48, 49, (52)

// (0), 3, 4, 7, 10, 11, 14, 17, 20, 23, 25, 28, 31, 34, 35, 38, 39, 42, 45,
// 46, 48, 49, (52)

// (0), 3, 4, 7, 10, 11, 14, 17, 20, 23, 25, 28, 31, 34, 35, 38, 39, 42, 45,
// 46, 49, (52)

// (0), 3, 4, 7, 10, 11, 14, 17, 20, 23, 25, 28, 31, 34, 35, 38, 39, 42, 45,
// 47, 48, 49, (52)

// (0), 3, 4, 7, 10, 11, 14, 17, 20, 23, 25, 28, 31, 34, 35, 38, 39, 42, 45,
// 47, 49, (52)

// (0), 3, 4, 7, 10, 11, 14, 17, 20, 23, 25, 28, 31, 34, 35, 38, 39, 42, 45,
// 48, 49, (52)

// In total, this set of adapters can connect the charging outlet to your device in 19208 distinct arrangements.

// You glance back down at your bag and try to remember why you brought so many adapters; there must be more than a trillion valid ways to arrange them! Surely, there must be an efficient way to count the arrangements.

// What is the total number of distinct ways you can arrange the adapters to connect the charging outlet to your device?

function analyzeAdapters (part, inputStr) {
  const inputArr = inputStr.split('\n');

  const adapters = inputArr.map(n => +n).sort((a, b) => a - b);         // in part 1 we must use all adapters, which must always increase in joltage, so there is only one possible sequence
                                                                        // in part 2, the hash is dynamically built from lower values, so again, we need the data to be sorted
  const device = adapters[adapters.length - 1] + 3;                     // the device itself is 3 higher than the highest adapter. figure out what its value is...
  adapters.push(device);                                                // ...and add the device itself to the list

  if (part === 1) {                                                     // PART 1: DETERMINE THE JOLTAGE OF THE CONNECTIONS BETWEEN THE SORTED ADAPTERS

    let diff1 = 0;
    let diff3 = 0;
    for (let i = 0; i < adapters.length; ++i) {
      const diff = !i ? adapters[i] : adapters[i] - adapters[i - 1];    // the first adapter always connects to the outlet which has joltage value of 0
      if (diff < 1 || diff > 3) throw "INVALID: NO SOLUTION";           // we are given that there must be a solution (thus diff must be 1, 2, or 3)
      if (diff === 1) ++diff1;
      if (diff === 3) ++diff3;
    }
    return diff1 * diff3;

  } else {                                                              // PART 2: COUNT THE NUMBER OF WAYS YOU CAN CONNECT LOWER ADAPTERS UP TO THE ADAPTER OF A GIVEN JOLTAGE

    const hash = {};                                                    // if an adapter has joltage i, then hash[i] should be the number of arrangements to get up to that device
    hash[0] = 1;                                                        // there is, of course, only one arrangement to get up to just the outlet
    for (const adapter of adapters) {                                   // for each adapter, in increasing order, if the joltage value of the adapter is X...
      hash[adapter] = (hash[adapter - 1] || 0)                          // ...then the number of ways to reach it is the sum of the number of ways to reach X-1 (if such adapter exists)...
        + (hash[adapter - 2] || 0)                                      // ...plus the number of ways to reach X-2 (if such adapter exists)...
        + (hash[adapter - 3] || 0);                                     // ...plus the number of ways to reach X-3 (if such adapter exists). to get to X, you must come from X-1, X-2, or X-3.
    }
    return hash[device];                                                // recall that device is the joltage value of your device. simply return the corresponding value saved in the hash

  }
}

// TEST CASES

const test = require('./_test');
const testNum = [1];
let input, expected;
const func = analyzeAdapters;
const sortedFunc = (...args) => func(...args).sort();                   // used when the order of the output does not matter
const modFunc = (...args) => func(...args) % 1000000007;                // used when the output is very large
const lowestTest = 0 || 0;
const highestTest = 0 || Infinity;

const sampleInputA = `16
10
15
5
1
11
7
19
6
12
4`;

const sampleInputB = `28
33
18
42
31
14
46
20
48
47
24
23
49
45
19
38
39
11
1
32
25
35
8
17
7
9
4
2
34
10
3`;

const actualInput = `104
83
142
123
87
48
102
159
122
69
127
151
147
64
152
90
117
132
63
109
27
47
7
52
59
11
161
12
148
155
129
10
135
17
153
96
3
93
82
55
34
65
89
126
19
72
20
38
103
146
14
105
53
77
120
39
46
24
139
95
140
33
21
84
56
1
32
31
28
4
73
128
49
18
62
81
66
121
54
160
158
138
94
43
2
114
111
110
78
13
99
108
141
40
25
154
26
35
88
76
145`;

// Test case 1
input = {
  part: 1,
  inputStr: sampleInputA,
};
expected = 7 * 5;
test(func, input, expected, testNum, lowestTest, highestTest);

// Test case 2
input = {
  part: 1,
  inputStr: sampleInputB,
};
expected = 22 * 10;
test(func, input, expected, testNum, lowestTest, highestTest);

// Test case 3
input = {
  part: 1,
  inputStr: actualInput,
};
expected = 2201;
test(func, input, expected, testNum, lowestTest, highestTest);

// Test case 4
input = {
  part: 2,
  inputStr: sampleInputA,
};
expected = 8;
test(func, input, expected, testNum, lowestTest, highestTest);

// Test case 5
input = {
  part: 2,
  inputStr: sampleInputB,
};
expected = 19208;
test(func, input, expected, testNum, lowestTest, highestTest);

// Test case 6
input = {
  part: 2,
  inputStr: actualInput,
};
expected = 169255295254528;
test(func, input, expected, testNum, lowestTest, highestTest);

File: 2025/day10.py
"""

--- Day 10: Factory ---

Just across the hall, you find a large factory. Fortunately, the Elves here have plenty of time to decorate. Unfortunately, it's because the factory machines are all offline, and none of the Elves can figure out the initialization procedure.

The Elves do have the manual for the machines, but the section detailing the initialization procedure was eaten by a Shiba Inu. All that remains of the manual are some indicator light diagrams, button wiring schematics, and joltage requirements for each machine.

For example:

[.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
[...#.] (0,2,3,4) (2,3) (0,4) (0,1,2) (1,2,3,4) {7,5,12,7,2}
[.###.#] (0,1,2,3,4) (0,3,4) (0,1,2,4,5) (1,2) {10,11,11,5,10,5}
The manual describes one machine per line. Each line contains a single indicator light diagram in [square brackets], one or more button wiring schematics in (parentheses), and joltage requirements in {curly braces}.

To start a machine, its indicator lights must match those shown in the diagram, where . means off and # means on. The machine has the number of indicator lights shown, but its indicator lights are all initially off.

So, an indicator light diagram like [.##.] means that the machine has four indicator lights which are initially off and that the goal is to simultaneously configure the first light to be off, the second light to be on, the third to be on, and the fourth to be off.

You can toggle the state of indicator lights by pushing any of the listed buttons. Each button lists which indicator lights it toggles, where 0 means the first light, 1 means the second light, and so on. When you push a button, each listed indicator light either turns on (if it was off) or turns off (if it was on). You have to push each button an integer number of times; there's no such thing as "0.5 presses" (nor can you push a button a negative number of times).

So, a button wiring schematic like (0,3,4) means that each time you push that button, the first, fourth, and fifth indicator lights would all toggle between on and off. If the indicator lights were [#.....], pushing the button would change them to be [...##.] instead.

Because none of the machines are running, the joltage requirements are irrelevant and can be safely ignored.

You can push each button as many times as you like. However, to save on time, you will need to determine the fewest total presses required to correctly configure all indicator lights for all machines in your list.

There are a few ways to correctly configure the first machine:

[.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
You could press the first three buttons once each, a total of 3 button presses.
You could press (1,3) once, (2,3) once, and (0,1) twice, a total of 4 button presses.
You could press all of the buttons except (1,3) once each, a total of 5 button presses.
However, the fewest button presses required is 2. One way to do this is by pressing the last two buttons ((0,2) and (0,1)) once each.

The second machine can be configured with as few as 3 button presses:

[...#.] (0,2,3,4) (2,3) (0,4) (0,1,2) (1,2,3,4) {7,5,12,7,2}
One way to achieve this is by pressing the last three buttons ((0,4), (0,1,2), and (1,2,3,4)) once each.

The third machine has a total of six indicator lights that need to be configured correctly:

[.###.#] (0,1,2,3,4) (0,3,4) (0,1,2,4,5) (1,2) {10,11,11,5,10,5}
The fewest presses required to correctly configure it is 2; one way to do this is by pressing buttons (0,3,4) and (0,1,2,4,5) once each.

So, the fewest button presses required to correctly configure the indicator lights on all of the machines is 2 + 3 + 2 = 7.

Analyze each machine's indicator light diagram and button wiring schematics. What is the fewest button presses required to correctly configure the indicator lights on all of the machines?


--- Part Two ---

All of the machines are starting to come online! Now, it's time to worry about the joltage requirements.

Each machine needs to be configured to exactly the specified joltage levels to function properly. Below the buttons on each machine is a big lever that you can use to switch the buttons from configuring the indicator lights to increasing the joltage levels. (Ignore the indicator light diagrams.)

The machines each have a set of numeric counters tracking its joltage levels, one counter per joltage requirement. The counters are all initially set to zero.

So, joltage requirements like {3,5,4,7} mean that the machine has four counters which are initially 0 and that the goal is to simultaneously configure the first counter to be 3, the second counter to be 5, the third to be 4, and the fourth to be 7.

The button wiring schematics are still relevant: in this new joltage configuration mode, each button now indicates which counters it affects, where 0 means the first counter, 1 means the second counter, and so on. When you push a button, each listed counter is increased by 1.

So, a button wiring schematic like (1,3) means that each time you push that button, the second and fourth counters would each increase by 1. If the current joltage levels were {0,1,2,3}, pushing the button would change them to be {0,2,2,4}.

You can push each button as many times as you like. However, your finger is getting sore from all the button pushing, and so you will need to determine the fewest total presses required to correctly configure each machine's joltage level counters to match the specified joltage requirements.

Consider again the example from before:

[.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
[...#.] (0,2,3,4) (2,3) (0,4) (0,1,2) (1,2,3,4) {7,5,12,7,2}
[.###.#] (0,1,2,3,4) (0,3,4) (0,1,2,4,5) (1,2) {10,11,11,5,10,5}
Configuring the first machine's counters requires a minimum of 10 button presses. One way to do this is by pressing (3) once, (1,3) three times, (2,3) three times, (0,2) once, and (0,1) twice.

Configuring the second machine's counters requires a minimum of 12 button presses. One way to do this is by pressing (0,2,3,4) twice, (2,3) five times, and (0,1,2) five times.

Configuring the third machine's counters requires a minimum of 11 button presses. One way to do this is by pressing (0,1,2,3,4) five times, (0,1,2,4,5) five times, and (1,2) once.

So, the fewest button presses required to correctly configure the joltage level counters on all of the machines is 10 + 12 + 11 = 33.

Analyze each machine's joltage requirements and button wiring schematics. What is the fewest button presses required to correctly configure the joltage level counters on all of the machines?

"""

# LIBRARIES
import time
import pathlib
import os
from math import *
from functools import *
from sys import *
from collections import *
from queue import *

# MODULES
from _test import test

# OPTIONAL VARIABLES
DISPLAY_EXTRA_INFO = True
# DISPLAY_EXTRA_INFO = False

def solve_linear_programming_problem(part, input_str, DEBUG = False, *args):

  # DATA STRUCTURES

  MACHINES = []


  # PARSE INPUT DATA

  input_arr = input_str.split('\n')
  ON = '#'

  for line in input_arr:
    components = line.split(' ')
    goal = components[0]
    buttons = components[1:-1]
    joltage = components[-1]

    MACHINES.append({
      'goal': [ c == ON for c in goal[1:-1] ],
      'buttons': [ [ int(n) for n in data[1:-1].split(',') ] for data in buttons ],
      'joltage_goal': [ int(n) for n in joltage[1:-1].split(',') ],
    })

  
  # ANALYZE

  TIME_AT_START = time.time()

  if part == 1:                                                                         # PART 1: THE CONSTRAINTS ON THE SYSTEM OF EQUATIONS ARE RELAXED;
                                                                                        # EACH 'EQUATION' MAPS TO JUST ODD/EVEN, NOT A SPECIFIC CONSTANT

    """
    We tackle this with recursion and it just works.
    """
  
    output = 0

    for machine_idx in range(len(MACHINES)):

      # RECURSIVE FUNCTION
      def go(button_idx, state, buttons_pressed):

        # BASE CASE
        if button_idx == len(MACHINES[machine_idx]['buttons']):

          return (state == MACHINES[machine_idx]['goal'], buttons_pressed)

        # RECURSIVE CASE
        else:

          CANDIDATES = []                                                               # Candidate states of buttons pressed that ultimately lead to goal state

          # Don't press current button, and recurse
          (res_if_no_press, buttons_pressed_if_no_press) = go(button_idx + 1, state, buttons_pressed)
          if res_if_no_press:
            CANDIDATES.append((True, buttons_pressed_if_no_press))

          # Press current button, and recurse
          state_copy = state.copy()
          for i in MACHINES[machine_idx]['buttons'][button_idx]:
            state_copy[i] = not state_copy[i]
          buttons_pressed_copy = buttons_pressed.copy()
          buttons_pressed_copy.append(button_idx)
          (res_if_press, buttons_pressed_if_press) = go(button_idx + 1, state_copy, buttons_pressed_copy)
          if res_if_press:
            CANDIDATES.append((True, buttons_pressed_if_press))

          # Only propagate the solution with the least number of buttons presesd, if there are multiple solutions
          CANDIDATES.sort(key=lambda candidate: len(candidate[1]))
          return (True, CANDIDATES[0][1]) if CANDIDATES else (False, [])

      # KICK-START RECURSION
      initial_state = [False] * len(MACHINES[machine_idx]['goal'])
      initial_buttons_pressed = []
      (res, buttons_pressed) = go(0, initial_state, initial_buttons_pressed)

      assert res, f'ERROR: No solution for machine_idx={machine_idx}'
      output += len(buttons_pressed)

    return output

  else:                                                                                 # PART 2: NOW, EACH EQUATION IN THE SYSTEM MAPS TO A SPECIFIC CONSTANT

    """
    I'm basically doing a brute force here. It works for the sample input, but takes way too long for the real data.
    Next, I'll end up doing a solution involving the `pulp` library, and apply it to both sample input and real input.
    """

    # if DEBUG:

    #   def serialize(times_buttons_pressed):
    #     return ''.join([ str(n) for n in times_buttons_pressed ])

    #   MEMO = {}
    #   def go(machine_idx, button_idx, times_buttons_pressed, joltage_state):

    #     joltage_goal = MACHINES[machine_idx]['joltage_goal']

    #     serialized_times_button_pressed = serialize(times_buttons_pressed)

    #     if (machine_idx, button_idx, serialized_times_button_pressed) not in MEMO:

    #       # Prune
    #       prune = False
    #       for i in range(len(joltage_state)):
    #         if joltage_state[i] > joltage_goal[i]:
    #           MEMO[(machine_idx, button_idx, serialized_times_button_pressed)] = (False, [], [])
    #           prune = True

    #       if not prune:

    #         if button_idx == len(times_buttons_pressed):

    #           MEMO[(machine_idx, button_idx, serialized_times_button_pressed)] = (
    #             joltage_state == joltage_goal,
    #             times_buttons_pressed,
    #             joltage_state,
    #           )

    #         else:

    #           CANDIDATES = []

    #           # res if don't press
    #           (res_if_no_press, buttons_pressed_if_no_press, joltage_state_if_no_press) = go(machine_idx, button_idx + 1, times_buttons_pressed, joltage_state)
    #           if res_if_no_press and joltage_state_if_no_press == joltage_goal:
    #             CANDIDATES.append((True, buttons_pressed_if_no_press, joltage_state_if_no_press))

    #           # res if press

    #           # prune by finding the lowest candidate number for max_presses
    #           max_presses = float('inf')
    #           for b_idx in MACHINES[machine_idx]['buttons'][button_idx]:
    #             delta = joltage_goal[b_idx] - joltage_state[b_idx]
    #             max_presses = min(max_presses, delta)

    #           for num_presses in range(1, max_presses + 1):
    #             times_buttons_pressed_copy = times_buttons_pressed.copy()
    #             times_buttons_pressed_copy[button_idx] += num_presses
    #             joltage_state_copy = joltage_state.copy()
    #             for i in MACHINES[machine_idx]['buttons'][button_idx]:
    #               joltage_state_copy[i] += num_presses
    #             (res_if_press, buttons_pressed_if_press, joltage_state_if_press) = go(machine_idx, button_idx + 1, times_buttons_pressed_copy, joltage_state_copy)
    #             if res_if_press and joltage_state_if_press == joltage_goal:
    #               CANDIDATES.append((True, buttons_pressed_if_press, joltage_state_if_press))

    #           if not CANDIDATES:
    #             # return (False, [], [])
    #             MEMO[(machine_idx, button_idx, serialized_times_button_pressed)] = (False, [], [])

    #           else:
    #             CANDIDATES.sort(key=lambda candidate: sum(candidate[1]))
    #             winning_candidate = CANDIDATES[0]
    #             # return winning_candidate
    #             MEMO[(machine_idx, button_idx, serialized_times_button_pressed)] = winning_candidate
          
    #     return MEMO[(machine_idx, button_idx, serialized_times_button_pressed)]

    #   print('RUNNING PART 2 ANALYSIS (PLEASE WAIT)...')

    #   print(f'Need to solve for {len(MACHINES)} machines')

    #   output = 0

    #   for machine_idx in range(len(MACHINES)):
    #     (res, buttons_pressed, _) = go(machine_idx, 0, [0] * len(MACHINES[machine_idx]['buttons']), [0] * len(MACHINES[machine_idx]['goal']))

    #     assert res, f'ERROR: No solution for machine_idx={machine_idx}'

    #     output += sum(buttons_pressed)

    #     print(f'Solved for machine {machine_idx}')

    #   print(f"(RUN TOOK {(time.time() - TIME_AT_START)} SECS)")
    #   return output

    """
    Recognizing that we basically have a system of linear equations, I asked ChatGPT how I might solve such a system
    in Python (without any reference to the AoC problem itself).

    It showed me an implementation using the `pulp` library, which helps with linear programming.
    """

    import pulp

    output = 0

    for machine_idx in range(len(MACHINES)):
      
      # STEP 1: Set up relationships from input data
      # An equation is something like 'a + b + f = 30', which means the number of times buttons 0, 1, and 5 are pressed collectively equals 30

      EQUATIONS = [{
        'LS': [],                                                                       # we will populate this with all buttons that might affect this joltage goal
        'RS': MACHINES[machine_idx]['joltage_goal'][joltage_goal_idx],                  # total button presses among the aforementioned group of buttons
      } for joltage_goal_idx in range(len(MACHINES[machine_idx]['joltage_goal']))]      # the above happens for each joltage goal
      
      # Iterate through each button, and look to see which joltage goals it affects
      for button_idx in range(len(MACHINES[machine_idx]['buttons'])):
        equation_list = MACHINES[machine_idx]['buttons'][button_idx]
        for joltage_goal_idx in equation_list:
          EQUATIONS[joltage_goal_idx]['LS'].append(button_idx)


      # STEP 2: Set up PuLP - Started with hard-coded implementation by ChatGPT, then I made it dynamic

      prob = pulp.LpProblem('MinSum', pulp.LpMinimize)                                  # some init stuff

      # Variables (nonnegative integers)
      x = pulp.LpVariable.dicts(                                                        # more init stuff
        "x",
        range(len(MACHINES[machine_idx]['buttons'])),
        lowBound=0,
        cat=pulp.LpInteger
      )

      # Objective: minimize total
      prob += pulp.lpSum(x[v] for v in x)                                               # more init stuff


      # Constraints

      # Hard coded example for 5 equations:
      # prob += x[0] + x[2] + x[3] == 3                                                 # this is like a + c + d = 3
      # prob += x[2] + x[3] + x[4] == 6                                                 # c + d + e = 6, etc.
      # prob += x[1] + x[3] == 1
      # prob += x[2] + x[4] == 5
      # prob += x[3] == 1
      for equation in EQUATIONS:
        # Here we are setting up a more dynamic version of the hard coded example lines above:
        # prob += (x[i0] + x[i1] + ... x[ik]) == RS
        prob += reduce(lambda acc, button_idx: acc + x[button_idx], equation['LS'], 0) == equation['RS']

      # Solve
      prob.solve(pulp.PULP_CBC_CMD(msg=False))
      solution = {v.name: int(v.value()) for v in prob.variables()}

      total_buttons_pressed = sum(solution.values())
      output += total_buttons_pressed

      if DEBUG and DISPLAY_EXTRA_INFO:
        print(f'===== SOLUTION FOR MACHINE {machine_idx} =====')
        for button_idx in sorted(solution.keys(), key=lambda button_idx: int(button_idx[2:])):
          if solution[button_idx] > 0:
            print(f'Button {button_idx[2:]} {tuple(MACHINES[machine_idx]['buttons'][int(button_idx[2:])])}: {solution[button_idx]} presses')
        print(f'Total: {total_buttons_pressed} button presses')
        print('')

    print(f"(RUN TOOK {(time.time() - TIME_AT_START)} SECS)")
    return output



# TEST CASES

test_num = [1]
test_input = None
test_expected = None
func = solve_linear_programming_problem
skipped_tests = set([ 2, 3, 4 ])
skipped_tests = set([ 3, 4 ])
skipped_tests = set([ 4 ])
skipped_tests = set([  ])
lowest_test = 0
highest_test = 0

FILENAME = pathlib.Path(__file__).name
FILENAME_SPLIT = FILENAME.split('.py')[0].split('day')
DAY_NUM = FILENAME_SPLIT[1] if len(FILENAME_SPLIT) > 1 else None
CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
INPUT_PATH = CURRENT_DIR + '\\' + (f"day{str(DAY_NUM).zfill(2)}-input.txt" if DAY_NUM else 'template-input.txt')
actual_input = open(INPUT_PATH, 'r').read()

sample_input = """[.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
[...#.] (0,2,3,4) (2,3) (0,4) (0,1,2) (1,2,3,4) {7,5,12,7,2}
[.###.#] (0,1,2,3,4) (0,3,4) (0,1,2,4,5) (1,2) {10,11,11,5,10,5}"""

# Test case 1
test_input = {
  'part': 1,
  'input_str': sample_input,
  'DEBUG': True,
}
test_expected = 2 + 3 + 2
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 2
test_input = {
  'part': 1,
  'input_str': actual_input,
  'DEBUG': False,
}
test_expected = 444
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 3
test_input = {
  'part': 2,
  'input_str': sample_input,
  'DEBUG': True,
}
test_expected = 33
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 4
test_input = {
  'part': 2,
  'input_str': actual_input,
  'DEBUG': False,
}
test_expected = 16513
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)